src/pmi/src/pmix.c

```diff
@@ -0,0 +1,558 @@
+
+    const char *jid;
+    pmi_errno = PMIU_msg_get_response_kvsname(&pmicmd, &jid);
+    PMIU_ERR_POP(pmi_errno);
+
+    MPL_strncpy(PMIx_proc.nspace, jid, PMIX_MAX_NSLEN + 1);
+    PMIU_Set_rank_kvsname(PMI_rank, jid);
+
+    if (!PMI_initialized) {
+        PMI_initialized = NORMAL_INIT_WITH_PM;
+    }
+
+    *proc = PMIx_proc;
+
+    /* TODO: add reference counting? */
+
+  fn_exit:
+    PMIU_cmd_free_buf(&pmicmd);
+    return pmi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+pmix_status_t PMIx_Finalize(const pmix_info_t info[], size_t ninfo)
+{
+    int pmi_errno = PMIX_SUCCESS;
+
+    struct PMIU_cmd pmicmd;
+    PMIU_cmd_init_zero(&pmicmd);
+
+    if (PMI_initialized > SINGLETON_INIT_BUT_NO_PM) {
+        PMIU_msg_set_query(&pmicmd, USE_WIRE_VER, PMIU_CMD_FINALIZE, no_static);
+
+        pmi_errno = PMIU_cmd_get_response(PMI_fd, &pmicmd);
+        PMIU_ERR_POP(pmi_errno);
+
+        shutdown(PMI_fd, SHUT_RDWR);
+        close(PMI_fd);
+    }
+
+  fn_exit:
+    PMIU_cmd_free_buf(&pmicmd);
+    return pmi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+pmix_status_t PMIx_Abort(int status, const char msg[], pmix_proc_t procs[], size_t nprocs)
+{
+    PMIU_printf(1, "aborting job:\n%s\n", msg);
+
+    struct PMIU_cmd pmicmd;
+    PMIU_msg_set_query_abort(&pmicmd, USE_WIRE_VER, no_static, status, msg);
+
+    /* ignoring return code, because we're exiting anyway */
+    PMIU_cmd_send(PMI_fd, &pmicmd);
+
+    PMIU_Exit(status);
+    return PMIX_SUCCESS;
+}
+
+pmix_status_t PMIx_Put(pmix_scope_t scope, const char key[], pmix_value_t * val)
+{
+    int pmi_errno = PMIX_SUCCESS;
+
+    /* This is a special hack to support singleton initialization */
+    if (PMI_initialized == SINGLETON_INIT_BUT_NO_PM) {
+        if (cached_singinit_inuse)
+            return PMIX_ERROR;
+        cached_singinit_key = MPL_strdup(key);
+        /* FIXME: save copy of value */
+        cached_singinit_inuse = true;
+        return PMIX_SUCCESS;
+    }
+
+    struct PMIU_cmd pmicmd;
+    PMIU_cmd_init_zero(&pmicmd);
+
+    if (val->type == PMIX_STRING) {
+        char *value = MPL_malloc(strlen(val->data.string) + 3, MPL_MEM_OTHER);
+        strcpy(value, "3,");
+        strcpy(value + 2, val->data.string);
+        PMIU_msg_set_query_kvsput(&pmicmd, USE_WIRE_VER, no_static, key, value);
+
+        pmi_errno = PMIU_cmd_get_response(PMI_fd, &pmicmd);
+        PMIU_ERR_POP(pmi_errno);
+
+        MPL_free(value);
+    } else if (val->type == PMIX_BYTE_OBJECT) {
+        /* encode the binary value and prepend type */
+        char *encoded_value = MPL_malloc(val->data.bo.size * 2 + 4, MPL_MEM_OTHER);
+        strcpy(encoded_value, "27,");
+        MPL_hex_encode(val->data.bo.size, val->data.bo.bytes, encoded_value + 3);
+
+        PMIU_msg_set_query_kvsput(&pmicmd, USE_WIRE_VER, no_static, key, encoded_value);
+
+        pmi_errno = PMIU_cmd_get_response(PMI_fd, &pmicmd);
+        PMIU_ERR_POP(pmi_errno);
+
+        MPL_free(encoded_value);
+    }
```

Comment:
Missing handling of other types. 

I suggest use `MPL_hex_encode` to encode the entire `val` always, and always decode at `kvsget`. Alternatively, can we wrap this block into `pmix_val_encode` and later block into `pmix_val_decode`. Effectively, you are doing your ad-hoc encode/decode algorithms. So with wrapping into the utility functions, we can modify or replace in the future.

Reply:
Sure, can move this to static function. Encoding even the string values should be OK. All the other types will be a pain but doable with some effort.

Reply:
last commit addresses this. can be squashed before merge.